File: 2025/572. Subtree of Another Tree/solution2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:

        def serialize(root: Optional[TreeNode]) -> str:
            if root:
                s = (f"#{str(root.val)}" 
                    f"{serialize(root.left)}" 
                    f"{serialize(root.right)}")
                return s
            else:
                return "#nil"

        return serialize(subRoot) in serialize(root)

File: 2025/572. Subtree of Another Tree/Solution2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        return serialize(root).contains(serialize(subRoot));
    }

    private String serialize(TreeNode root) {
        if (root != null) {
            return "#" + root.val + serialize(root.left) + serialize(root.right);
        } else {
            return "#nil";
        }
    }
}